rust/actyx/ax-aql/src/language/types.rs

```diff
@@ -8,17 +8,258 @@ use crate::{
+                        fields.into_iter().collect::<Vec<_>>().try_into().unwrap(),
+                    )))
+                }
+            }
+            _ => Ok(self),
+        }
+    }
+
+    /// Calculate supertype without collapsing
+    fn is_supertype_of(&self, other: &Type) -> bool {
+        match (self, other) {
+            (Type::Atom(a), Type::Atom(b)) => match (a, b) {
+                (TypeAtom::Bool(None), TypeAtom::Bool(Some(_))) => true,
+                (TypeAtom::Number(None), TypeAtom::Number(Some(_))) => true,
+                (TypeAtom::String(None), TypeAtom::String(Some(_))) => true,
+                (TypeAtom::Record(a), TypeAtom::Record(b)) => {
+                    let a = a.iter().collect::<BTreeSet<_>>();
+                    let b = b.iter().collect::<BTreeSet<_>>();
+
+                    a.intersection(&b).cloned().collect::<BTreeSet<_>>() == b
+                }
+                (TypeAtom::Universal, _) => true,
+                _ => false,
+            },
+            (Type::Union(a), Type::Union(b)) => {
+                let a = Type::flatten_union(a.as_ref());
+                let b = Type::flatten_union(b.as_ref());
+
+                a.intersection(&b).cloned().collect::<BTreeSet<_>>() == b
+            }
+            (Type::Union(union), non_union) => Type::flatten_union(union.as_ref()).contains(non_union),
+            (non_union, Type::Union(union)) => Type::flatten_union(union.as_ref()).contains(non_union),
+            (Type::Array(a), Type::Array(b)) => a.is_supertype_of(b.as_ref()),
+            (Type::Dict(a), Type::Dict(b)) => a.as_ref().is_supertype_of(b.as_ref()),
+            _ => false,
+        }
+    }
+
+    /// Flatten union tree into a set of type references without collapsing
+    fn flatten_union((a, b): &(Type, Type)) -> BTreeSet<&Type> {
+        let mut under_types_vec = vec![a, b];
+
+        loop {
+            let mut non_unions = vec![];
+            let mut unions = vec![];
+
+            under_types_vec.split_off(0).into_iter().for_each(|ty| match ty {
+                Type::Union(pair) => unions.push(pair),
+                _ => non_unions.push(ty),
+            });
+            std::mem::swap(&mut under_types_vec, &mut non_unions);
+
+            if unions.is_empty() {
+                break;
+            }
+
+            under_types_vec.extend(unions.into_iter().flat_map(|x| {
+                let (a, b) = x.as_ref();
+                [a, b]
+            }));
+        }
+
+        under_types_vec.into_iter().collect()
+    }
+
+    /// Flatten union tree into a set of types with collapsing
+    fn flatten_union_collapsing((a, b): &(Type, Type)) -> Result<BTreeSet<Type>, CollapseError> {
+        let mut under_types_vec = vec![a.clone(), b.clone()];
+        let mut collapse_errors = vec![];
+
+        loop {
+            let mut non_unions = vec![];
+            let mut unions = vec![];
+
+            under_types_vec
+                .split_off(0)
```

Comment:
Isn't this a no-op?

Reply:
the goal is to take all members of the vec and leave it empty.
i should replace this with `std::mem::take(&mut under_types_vec)` tho